packages/transport/protocol-v2-e2e.ts

```diff
@@ -0,0 +1,538 @@
+        });
+
+        if (!cpaceTrezor.success) throw new Error('cpaceTrezor unsuccessful');
+
+        // 1. Set *shared_secret* = X25519(*cpace_host_private_key*, *cpace_trezor_public_key*).
+        // 2. Set *tag* = SHA-256(*shared_secret*).
+
+        const tag = protocolThp.getShareSecret(
+            Buffer.from(cpaceTrezor.payload.message.cpace_trezor_public_key, 'hex'),
+            hostKeys.privateKey,
+        );
+
+        const codeEntry = await call('ThpCodeEntryTag', {
+            tag,
+            // tag: Buffer.from(
+            //     '2278a1ba6b91e6d5faaecf0978340b55ab4c74824f0d45222cdc3d4a7395fc65',
+            //     'hex',
+            // ),
+        });
+
+        if (!codeEntry.success) {
+            throw new Error('ThpCodeEntryTag unsuccessful');
+        }
+
+        const { secret } = codeEntry.payload.message;
+        const sha = sha256(Buffer.from(secret, 'hex'));
+
+        console.warn('secret compare with commitment', sha.compare(handshakeCommitment));
+
+        return codeEntry;
+    };
+
+    const qrCodePairing = (pairingSecret: any, state: any) => {
+        const shaCtx = createHash('sha256');
+        shaCtx.update(state.handshakeCredentials.handshakeHash);
+        shaCtx.update(Buffer.from(pairingSecret, 'hex'));
+        shaCtx.update(Buffer.from('PairingMethod_QrCode', 'utf-8'));
+        const qrCode = shaCtx.digest().subarray(0, 16);
+        const tagSha = sha256(Buffer.from(qrCode));
+
+        return call('ThpQrCodeTag' as any, {
+            tag: tagSha,
+        });
+    };
+
+    const nfcPairing = (pairingSecret: any, state: any) => {
+        const shaCtx = createHash('sha256');
+        shaCtx.update(state.handshakeCredentials.handshakeHash);
+        shaCtx.update(Buffer.from(pairingSecret, 'hex'));
+        shaCtx.update(Buffer.from('PairingMethod_NfcUnidirectional', 'utf-8'));
+        const nfcCode = shaCtx.digest().subarray(0, 16);
+        const tagSha = sha256(Buffer.from(nfcCode));
+
+        return call('ThpNfcUnidirectionalTag', {
+            tag: tagSha,
+        });
+    };
+
+    const call = async (name: string, data: any): Promise<any> => {
+        const state = device.transports[0].channels[0];
+
+        console.warn('Call', name, state);
+
+        const result = await transport.call({
+            name,
+            data,
+            protocolState: state,
+            ...device.transports[0],
+        }).promise;
+
+        if (result.success) {
+            if (
+                (name === 'ThpStartPairingRequest' || name === 'ThpCodeEntryChallenge') &&
+                result.payload.type === 'ThpPairingPreparationsFinished'
+            ) {
+                console.warn('Waitin before sending ' + USE_PAIRING + ' tag');
+
+                await new Promise(resolve => setTimeout(resolve, 1000));
+
+                const debugState = await call('DebugLinkGetState', {});
+                const { thp_pairing_code_entry_code, thp_pairing_secret } =
```

Comment:
`code_qr_code` and `code_nfc_unidirectional` should be transmitted from Trezor to Host instead of `thp_pairing_secret`.